pkgs/applications/networking/browsers/firefox/common.nix

```diff
@@ -164,6 +166,20 @@ buildStdenv.mkDerivation ({
 
   inherit src unpackPhase meta;
 
+  # add profiling and a second configure phase if we build with pgo
```

Comment:
Why separate phases rather than splitting the build in two?

We need a regular firefox build anyways because pgo will never be reproducible. 

We could then use the unoptimised firefox to execute `profilingPhase` in a separate profiling derivation and then use that derivation's output to build Firefox again. Would that introduce a lot of complexity?

Reply:
For a regular firefox build you can just set `pgoSupport = false`, the change is quite clean in that regard. This was also the easiest way to implement this change as it reuses most everything and is itself reusable by other consumers.

> Would that introduce a lot of complexity?

Yes, needless complexity as everything is hidden behind the `pgoSupport` flag.

> We need a regular firefox build anyways because pgo will never be reproducible.

You're saying we should stick with a slower browser for reproducibility.'s sake, this make me sad and I don't know how to even respond.

Reply:
> For a regular firefox build you can just set `pgoSupport = false`, the change is quite clean in that regard.

What I mean is that we'd be building unoptimised firefox twice. Once for "reproducible" firefox and once for firefox-pgo.

Wouldn't it be better if we just took the regular firefox, generated profdata with it and then fed that profdata to the firefox drv again to build firefox-pgo?  
That way we cache the unoptimised build and it makes it a lot easier to inject your own profdata without rebuilding Firefox twice. This seems a lot cleaner to me architecturally.

> Yes, needless complexity as everything is hidden behind the `pgoSupport` flag.

That wouldn't change. The profiling would just happen in a separate drv rather than a phase. Same code, different place.

> You're saying we should stick with a slower browser for reproducibility.'s sake

No, I'm saying it should be there as an option. It'd be the one included in the graphical ISOs for example because the ISO's usage of Firefox is not performance critical while its integrity is critical. Users could obviously opt to use the optimised Firefox in their systems if they value performance above r13y. (I'd do that.)

> Why does firefox-bin´get a free pass?

It doesn't. From a r13y standpoint, binary packages are about as bad as it gets. They're 100% deterministic but completely fail on the transparency front.

Reply:
> Wouldn't it be better if we just took the regular firefox, generated profdata with it and then fed that profdata to the firefox drv again to build firefox-pgo?

No, that wouldn't work, because the first pass firefox isn't a standard build, but one that features instrumentation for profile generation.

Reply:
I see, thanks. That's what I wanted to know.

Reply:
As it is we're not reproducible anyway.

> error: derivation '/nix/store/a8h51kq58v2qn7jaki9iw28b05xk5x97-firefox-unwrapped-98.0.2.drv' may not be deterministic: output '/nix/store/5pyd9j5ag49cayg9rgdicmm1sgsknzxd-firefox-unwrapped-98.0.2-debug' differs

Reply:
Nah, I don't think firefox has been reproducible for a long time now. It's just that it's impossible to be reproducible with PGO while it should be possible without.

Reply:
> Nah, I don't think firefox has been reproducible for a long time now. It's just that it's impossible to be reproducible with PGO while it should be possible without.

* I would assume not possible for something as complex as a browser but I would imagine a deterministically running program (like no multithreading influence, rerendering, etc.) would be able to be reproducibly PGO optimizable.